File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
     unordered_map<char, int> seen; // freq map
     int l = 0;
     int maxFreq = 0;
     int windowSize = 0;
     int lettersToChange = 0;

     for(int r = 0; r < s.length(); r++)
     {
        seen[s[r]]++; // get the char freq
        maxFreq = max(maxFreq, seen[s[r]]); // get the highest char freq
        windowSize = r-l+1; // the window size (aka the distance between the right and left pointers)
        lettersToChange = windowSize-maxFreq; // calculate how many letters in the window need to be changed in order to have one distinct character

        if(lettersToChange > k) // if that number is > than the number of chars we're allowed to change
        {
            seen[s[l]]--; // shrink the window by removing the leftmost character
            l++; // move the left pointer up
        }

        windowSize = r-l+1; // update the windowSize assuming the window has shrank
     }

     return windowSize;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charSet = set()
        l = 0
        res = 0

        for r in range(len(s)):
            while s[r] in charSet: # while s[r] is already in the set aka a duplicate
                charSet.remove(s[l]) # remove the left character from the set
                l += 1 # shrink the window size
            charSet.add(s[r]) # add the original characters to the set
            res = max(res, r - l + 1) # calculate the longest length
        return res